app/src/main/java/org/connectbot/TerminalView.java

```diff
@@ -150,6 +171,282 @@ public TerminalView(Context context, TerminalBridge bridge) {
+		setTypeface(Typeface.MONOSPACE);
+		onFontSizeChanged(bridge.getFontSize());
+
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
+			setTextIsSelectable(true);
+
+			initSelectionCallback();
+
+			gestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
+				private TerminalBridge bridge = TerminalView.this.bridge;
+				private float totalY = 0;
+
+				@Override
+				public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
+					// if releasing then reset total scroll
+					if (e2.getAction() == MotionEvent.ACTION_UP) {
+						totalY = 0;
+					}
+
+					totalY += distanceY;
+					final int moved = (int) (totalY / bridge.charHeight);
+
+					if (moved != 0) {
+						int base = bridge.buffer.getWindowBase();
+						bridge.buffer.setWindowBase(base + moved);
+						totalY = 0;
+
+						refreshTextFromBuffer();
+					}
+
+					return true;
+				}
+
+				@Override
+				public boolean onSingleTapConfirmed(MotionEvent e) {
+					viewPager.performClick();
+					return super.onSingleTapConfirmed(e);
+				}
+			});
+		}
+	}
+
+	@TargetApi(11)
+	private void initSelectionCallback() {
+		this.setCustomSelectionActionModeCallback(new ActionMode.Callback() {
+			private static final int PASTE = 0;
+
+			@Override
+			public boolean onCreateActionMode(ActionMode mode, Menu menu) {
+				TerminalView.this.selectionActionMode = mode;
+
+				menu.add(0, PASTE, 2, "Paste")
+						.setIcon(R.drawable.ic_action_paste)
+						.setShowAsAction(MenuItem.SHOW_AS_ACTION_WITH_TEXT | MenuItem.SHOW_AS_ACTION_IF_ROOM);
+
+				return true;
+			}
+
+			@Override
+			public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
+				if (item.getItemId() == PASTE) {
+					String clip = clipboard.getText().toString();
+					TerminalView.this.bridge.injectString(clip);
+					mode.finish();
+					return true;
+				}
+
+				return false;
+			}
+
+			@Override
+			public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
+				return false;
+			}
+
+			@Override
+			public void onDestroyActionMode(ActionMode mode) {
+			}
+		});
+	}
+
+	public void copyCurrentSelectionToClipboard() {
+		ClipboardManager clipboard =
+				(ClipboardManager) TerminalView.this.context.getSystemService(Context.CLIPBOARD_SERVICE);
+		if (currentSelection.length() != 0) {
+			clipboard.setText(currentSelection);
+		}
+
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB &&
+			selectionActionMode != null) {
+			selectionActionMode.finish();
+			selectionActionMode = null;
+		}
+	}
+
+	@Override
+	protected void onSelectionChanged(int selStart, int selEnd) {
+		if (selStart <= selEnd) {
+			currentSelection = getText().toString().substring(selStart, selEnd);
+		}
+		super.onSelectionChanged(selStart, selEnd);
+	}
+
+	@Override
+	public boolean performLongClick() {
+		refreshTextFromBuffer();
+		return super.performLongClick();
+	}
+
+	@Override
+	public boolean onTouchEvent(MotionEvent event) {
```

Comment:
Also, I had problems with performClick() on M preview in my tests. I've just try on my Nexus 9 Marshmallow (release) and I can't select anything, so it seems the problem is still here.

I resolved it by using a TouchListener on TerminalView (cf https://github.com/alescdb/connectbot/commit/cfbbf3d51e7292d0da1c9fafe41d9948599244ad#diff-431cd3b89538f02edadfe658476256b6R161)

Nice work btw :-)